secp256k1-sys/src/lib.rs

```diff
@@ -673,60 +672,131 @@ mod fuzz_dummy {
+        assert!(rustsecp256k1_v0_4_1_context_preallocated_size(flags) + core::mem::size_of::<c_uint>() <= CTX_SIZE);
         CTX_SIZE
     }
 
-    static HAVE_PREALLOCATED_CONTEXT: AtomicUsize = AtomicUsize::new(0);
-    const HAVE_CONTEXT_NONE: usize = 0;
-    const HAVE_CONTEXT_WORKING: usize = 1;
-    const HAVE_CONTEXT_DONE: usize = 2;
+    static HAVE_PREALLOCATED_CONTEXT: once::Once = once::Once::INIT;
     static mut PREALLOCATED_CONTEXT: [u8; CTX_SIZE] = [0; CTX_SIZE];
+    #[cfg(feature = "std")]
+    thread_local! {
+        static CACHE_HAVE_PREALLOCATED_CONTEXT: std::cell::Cell<bool> = std::cell::Cell::new(false);
+    }
     pub unsafe fn secp256k1_context_preallocated_create(prealloc: *mut c_void, flags: c_uint) -> *mut Context {
         // While applications should generally avoid creating too many contexts, sometimes fuzzers
         // perform tasks repeatedly which real applications may only do rarely. Thus, we want to
         // avoid being overly slow here. We do so by having a static context and copying it into
         // new buffers instead of recalculating it. Because we shouldn't rely on std, we use a
         // simple hand-written OnceFlag built out of an atomic to gate the global static.
-        let mut have_ctx = HAVE_PREALLOCATED_CONTEXT.load(Ordering::Relaxed);
-        while have_ctx != HAVE_CONTEXT_DONE {
-            if have_ctx == HAVE_CONTEXT_NONE {
-                have_ctx = HAVE_PREALLOCATED_CONTEXT.swap(HAVE_CONTEXT_WORKING, Ordering::AcqRel);
-                if have_ctx == HAVE_CONTEXT_NONE {
-                    assert!(rustsecp256k1_v0_4_1_context_preallocated_size(SECP256K1_START_SIGN | SECP256K1_START_VERIFY) + std::mem::size_of::<c_uint>() <= CTX_SIZE);
-                    assert_eq!(rustsecp256k1_v0_4_1_context_preallocated_create(
-                            PREALLOCATED_CONTEXT[..].as_ptr() as *mut c_void,
-                            SECP256K1_START_SIGN | SECP256K1_START_VERIFY),
-                        PREALLOCATED_CONTEXT[..].as_ptr() as *mut Context);
-                    assert_eq!(HAVE_PREALLOCATED_CONTEXT.swap(HAVE_CONTEXT_DONE, Ordering::AcqRel),
-                        HAVE_CONTEXT_WORKING);
-                } else if have_ctx == HAVE_CONTEXT_DONE {
-                    // Another thread finished while we were swapping.
-                    HAVE_PREALLOCATED_CONTEXT.store(HAVE_CONTEXT_DONE, Ordering::Release);
-                }
-            } else {
-                // Another thread is building, just busy-loop until they're done.
-                assert_eq!(have_ctx, HAVE_CONTEXT_WORKING);
-                have_ctx = HAVE_PREALLOCATED_CONTEXT.load(Ordering::Acquire);
-                #[cfg(feature = "std")]
-                std::thread::yield_now();
-            }
+
+        #[cfg(feature = "std")]
+        let should_initialize = !CACHE_HAVE_PREALLOCATED_CONTEXT.with(|value| value.get());
```

Comment:
a) why not have the thread_local be inside the `once` instead of outside, so that any re-use gets the same benefit, but (b) if we make `once` conditionally switch to the `std` version if `std` is set, then we'd lose it anyway, so maybe makes sense to drop it? Or were you thinking its a nice optimization on top of `std::Once`?

Reply:
a) I don't see an easy and obvious way to do it. If multiple instances of `Once` are created they need to have different thread locals. The only way to achieve it seems to be having some hash map in thread local keyed by addresses from allocator. I don't think the overhead would help.
b) didn't occur to me but indeed it could be an optimization of std Once.

One more caveat is I'm not sure if the `with` method is fast enough to still be better than just using atomics for all accesses.

Reply:
a) does Rust not have instance field thread_locals?
b) right, the `with` method is probably a `Once` on the inside, so probably not something we want to do here :).

Reply:
a) I don't see how they could be tracked
b) It's not `Once` but looks like [on some platforms it may do a syscall](https://doc.rust-lang.org/src/std/thread/local.rs.html#166). :(

I think it's better to just revert it. If someone measures significant performance change we could feature gate it.

Reply:
I believe other languages just heap-allocate to implement instance thread locals.